src/lib.rs

```diff
@@ -369,10 +430,29 @@ pub unsafe extern "C" fn path_readlink(
     path_ptr: *const u8,
     path_len: usize,
     buf: *mut u8,
-    _buf_len: Size,
+    buf_len: Size,
     bufused: *mut Size,
 ) -> Errno {
-    unreachable()
+    let fd = wasi_filesystem::Descriptor::from_raw(fd as _);
+
+    let path = slice::from_raw_parts(path_ptr, path_len);
+
+    register_buffer(buf, buf_len);
+
+    let result = fd.readlink_at(path);
```

Comment:
I mentioned this in chat, but I think it's worth mentioning here as well. I don't think that the wasi-preview2 API here can be used to polyfill `path_readlink` since there's no way to call `path_readlink` and guarantee there will be no trap. The buffer length isn't known ahead of time and if the length is too small it'll kill the whole program in `cabi_realloc`.

Reply:
I just added another patch which allocates a `PATH_MAX`-sized buffer on the stack to fix this issue. I believe it's just `readlink` that needs to do this, so this might be ok.

Reply:
To what extent is that actually a limit of any system? Wasm presumably can't pick a `PATH_MAX` because every system is different, so is this something that the WASI spec and hosts will have to limit? For example if a Unix host supports a symlink target beyond `PATH_MAX` would it return some sort of error indicating that it can't be returned to the wasm module?

It looks like in `wasi-common` there's no checking of any `PATH_MAX` or such right now?

Reply:
Linux's `PATH_MAX` is 4096. BSD variants and macOS have 1024. Windows's `MAX_PATH` is 260. POSIX may have said that `PATH_MAX` might not be defined or might not be a constant, in practice, tons of C code in the world uses `char buf[PATH_MAX]` and it's unlikely that any practical Unix will break that for the foreseeable future.

WASI doesn't currently spec a 4096-byte limit, but it probably should, for the same reason.

Reply:
Regardless of existing C I always thought `PATH_MAX` was just some `#define` in libc that anyone can configure and it would also actually vary depending on the filesystem under the hood. I would naively expect that someone somewhere wanted a really long path, implemented support, and so long as you "do the right thing" then code which assumes `PATH_MAX` is enough would fail and "working" code would produce the right result.

Given https://serverfault.com/questions/9546/filename-length-limits-on-linux and my inability to locally at least do anything beyond 4096 characters I guess this seems fine for now.